caluma/form/historical_schema.py

```diff
@@ -190,13 +190,13 @@ def resolve_value(self, info, as_of, *args):
         answerdocuments.sort(key=lambda x: x.sort)
 
         documents = [
-            models.Document.history.filter(id=ad.document_id).filter(
-                history_date__lte=as_of
-            )[0]
-            for ad in answerdocuments
+            models.Document.history.filter(
+                id=ad.document_id, history_date__lte=as_of
+            ).latest("history_date")
+            for ad in set(answerdocuments)
         ]
 
-        return documents
+        return set(documents)
```

Comment:
You will lose sorting if you put it in a set, as a set is an unordered data structure.

If you want to preserve ordering, you'd need something like `return sorted(set(documents), key=...)`
instead

Reply:
Great find! It's fixed.